### DISCUSSION TITLE
Complex Numbers and Wavefunctions in MOOSE

### DISCUSSION THREAD

**Discussion by gj19866** [idaholab/moose#25512] [Q&A Modules: Phase field]
Hi there,

I am looking to utilize MOOSE to model the behavior of wavefunctions.
Is there any president of people doing this and any similar work that you could point me towards?

This will also involve the handling of complex numbers. What is the best way to handle complex numbers in MOOSE?

All the best,
Gillian J

**Comment by GiudGiud** [comment]
@cticenhour @salaudeen-ya and @aaelmeli  will have some insights on this

**Reply by gj19866** [reply]
Great thanks, I look forward to hearing from them!

**Comment by cticenhour** [comment]
Regarding complex numbers, we solve for complex quantities regularly in the [electromagnetics module](https://mooseframework.inl.gov/modules/electromagnetics/index.html), but we have to separate the real and imaginary components of the field variable into two field variables, as MOOSE doesn't perform the complex number calculation natively. This has worked well, but considering that complex number support exists in PETSc and libMesh, we do intend to add that support at the MOOSE level at some point (pending time and funding constraints). 

I do not have experience with quantum wavefunctions, so I will let others speak to that, but feel free to pose any other questions you have.

**Reply by gj19866** [reply]
Great, thank you!

**Reply by aaelmeli** [reply]
I do have some experience in modeling mechanical waves through moose. This includes acoustics, elastics and viscoelastic waves. As @cticenhour mentioned, you will need to decouple the equations to the real and imaginary and then model both fields as two (real) dependent field living on the same mesh. there are some examples on this [here](https://mooseframework.inl.gov/modules/tensor_mechanics/1d_elastic_waves.html).
